contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol

```diff
@@ -0,0 +1,1678 @@
+        ++_i;
+      }
+    }
+  }
+
+  /**
+    @notice
+    Pays out a split for a project's funding cycle configuration.
+  
+    @param _split The split to distribute payouts to.
+    @param _amount The total amount being distributed to the split, as a fixed point number with the same number of decimals as this terminal.
+    @param _feeDiscount The amount of discount to apply to the fee, out of the MAX_FEE.
+
+    @return netPayoutAmount The amount sent to the split after subtracting fees.
+  */
+  function _distributeToPayoutSplitsOf(
+    JBSplit memory _split,
+    uint256 _projectId,
+    uint256 _group,
+    uint256 _amount,
+    uint256 _feeDiscount
+  ) internal returns (uint256 netPayoutAmount) {
+    // If there's an allocator set, transfer to its `allocate` function.
+    if (_split.allocator != IJBSplitAllocator(address(0))) {
+      // If the split allocator is set as feeless, this distribution is not eligible for a fee.
+      if (
+        _feeDiscount == JBConstants.MAX_FEE_DISCOUNT || isFeelessAddress[address(_split.allocator)]
+      )
+        netPayoutAmount = _amount;
+        // This distribution is eligible for a fee since the funds are leaving this contract and the allocator isn't listed as feeless.
+      else {
+        unchecked {
+          netPayoutAmount = _amount - _feeAmount(_amount, fee, _feeDiscount);
+        }
+      }
+
+      // Trigger any inherited pre-transfer logic.
+      _beforeTransferTo(address(_split.allocator), netPayoutAmount);
+
+      // Create the data to send to the allocator.
+      JBSplitAllocationData memory _data = JBSplitAllocationData(
+        token,
+        netPayoutAmount,
+        decimals,
+        _projectId,
+        _group,
+        _split
+      );
+
+      // Trigger the allocator's `allocate` function.
+      // If this terminal's token is ETH, send it in msg.value.
+      bool _success;
+
+      if(ERC165Checker.supportsInterface(address(_split.allocator), type(IJBSplitAllocator).interfaceId))
+        try _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data) {
+          _success = true;
+        }
+        catch {}
+      
+      if(!_success) {
+        // Trigger any inhereted post-transfer cancelation logic.
+        _cancelTransferTo(address(_split.allocator), netPayoutAmount);
+
+        // Set the net payout amount to 0 to signal the reversion.
+        netPayoutAmount = 0;
+
+        // Add undistributed amount back to project's balance.
+        store.recordAddedBalanceFor(_projectId, _amount);
+      }
+
+      // Otherwise, if a project is specified, make a payment to it.
+    } else if (_split.projectId != 0) {
+      // Get a reference to the Juicebox terminal being used.
+      IJBPaymentTerminal _terminal = directory.primaryTerminalOf(_split.projectId, token);
+
+      // The project must have a terminal to send funds to.
+      if (_terminal == IJBPaymentTerminal(address(0))) revert TERMINAL_IN_SPLIT_ZERO_ADDRESS();
```

Comment:
Should we validate it's really a terminal (erc165) too?

Reply:
I was also thinking of this yesterday, only reason I had not to was the scope i.e whether this should be the responsibility of the project owner, but if we are doing it for allocator then you're right we should do it for terminal too

Reply:
nvm, when calling setPrimaryTerminal in the directory, a bunch of static calls are made to the terminal -> these would revert if it's not a valid terminal